scripts/gen_kobject_list.py

```diff
@@ -0,0 +1,443 @@
+            continue
+
+        ub = child.attributes["DW_AT_upper_bound"]
+        if not ub.form.startswith("DW_FORM_data"):
+            continue
+
+        size_found = True
+        elements = elements * (ub.value + 1)
+
+    if not size_found:
+        return
+
+    type_env[die.offset] = ArrayType(die.offset, elements, type_offset)
+
+
+def get_filename_lineno(die):
+    lp_header = die.dwarfinfo.line_program_for_CU(die.cu).header
+    files = lp_header["file_entry"]
+    includes = lp_header["include_directory"]
+
+    fileinfo = files[die.attributes["DW_AT_decl_file"].value - 1]
+    filename = fileinfo.name.decode("utf-8")
+    filedir = includes[fileinfo.dir_index - 1].decode("utf-8")
+
+    path = os.path.join(filedir, filename)
+    lineno = die.attributes["DW_AT_decl_line"].value
+    return (path, lineno)
+
+
+def find_kobjects(elf, syms):
+    if not elf.has_dwarf_info():
+        sys.stderr.write("ELF file has no DWARF information\n");
+        sys.exit(1)
+
+    kram_start = syms["__kernel_ram_start"]
+    kram_end = syms["__kernel_ram_end"]
+
+    di = elf.get_dwarf_info()
+
+    variables = []
+
+    # Step 1: collect all type information.
+    for CU in di.iter_CUs():
+        CU_path = CU.get_top_DIE().get_full_path()
+        lp = di.line_program_for_CU(CU)
+
+        for idx, die in enumerate(CU.iter_DIEs()):
+            # Unions are disregarded, kernel objects should never be union
+            # members since the memory is not dedicated to that object and
+            # could be something else
+            if die.tag == "DW_TAG_structure_type":
+                analyze_die_struct(die)
+            elif die.tag == "DW_TAG_array_type":
+                analyze_die_array(die)
+            elif die.tag == "DW_TAG_variable":
+                variables.append(die)
+
+    # Step 2: filter type_env to only contain kernel objects, or structs and
+    # arrays of kernel objects
+    bad_offsets = []
+    for offset, type_object in type_env.items():
+        if not type_object.has_kobject():
+            bad_offsets.append(offset)
+
+    for offset in bad_offsets:
+        del type_env[offset]
+
+    # Step 3: Now that we know all the types we are looking for, examine
+    # all variables
+    all_objs = []
+
+    # Gross hack, see below
+    work_q_found = False
+
+    for die in variables:
+        name = die_get_name(die)
+        if not name:
+            continue
+
+        type_offset = die_get_type_offset(die)
+
+        # Is this a kernel object, or a structure containing kernel objects?
+        if type_offset not in type_env:
+            continue
+
+        if "DW_AT_declaration" in die.attributes:
+            # FIXME: why does k_sys_work_q not resolve an address in the DWARF
+            # data??? Every single instance it finds is an extern definition
+            # but not the actual instance in system_work_q.c
+            # Is there something weird about how lib-y stuff is linked?
```

Comment:
Yes, actually: the kernel directory is always presented to the link as an already-linked static library, to prevent the case where pulling symbols from one file pulls in static data too and bloats the image.  It looks like k_sys_work_q is only ever referenced from inside the kernel, so that may be triggering some variant DWARF output.   How does this look in e.g. gdb?  Can you see the type there?

Reply:
I can't seem to find it in the DWARF output at all except as an extern reference. It's really baffling. I'm going to file a bug for further investigation.